### PR TITLE
Make Guided Missile Free (and Maintenance overhaul)

### PR DESCRIPTION
**Pull request by itanasi**
Adds support for NoMaintenance
Applies feature to Guided Missile
Change listed in #4697

### CODE REVIEW

**Comment by temurakami** [comment]
Wait, put on hold. Didn't add translation entries

**Comment by SomeTroglodyte** [comment]
No template necessary, the json will be picked up. But don't include unrelated build script changes.

You may get wrong results though as far as I can guess - they can be counted twice since they can also count as garrisoned? I'll leave looking more thoroughly to you.

**Comment by itanasi** [comment]
New version built.
This adds a more flexible and robust calculation for overlapping maintenance modifiers
And now Free Units will properly drop the most expensive units
Tested manually using print statements that the maintenance value gets updated and sorting/dropping works
Not sure about the float->double->float conversions, nor if there's a more efficient way than this iterative manner

File: core/src/com/unciv/logic/civilization/CivInfoStats.kt [hunk 1]
```diff
@@ -32,17 +32,38 @@ class CivInfoStats(val civInfo: CivilizationInfo) {
...
-            )
-            numberOfUnitsToPayFor += numberOfUnitsWithDiscount * unique.params[0].toFloat() / 100f
+        // Handle unit maintenance discounts
+        // Have to capture global and per-unit
+        // Free Garrison already removed above from sequence
+        // Initialize maintenance cost per unit, default 1 aka 100%
+        // Note all discounts are in the form of -X%, such as -25 for 25% reduction
+        for (unit in unitsToPayFor){
+            if (unit.hasUnique(UniqueType.UnitMaintenanceDiscount)){
+                unit.maintenance = 1f + unit.getMatchingUniques(UniqueType.UnitMaintenanceDiscount).first().params[0].toFloat()/100f
```
Comment (yairm210) [inline core/src/com/unciv/logic/civilization/CivInfoStats.kt:42]: Use unique.param[0].toPercent()

File: core/src/com/unciv/logic/civilization/CivInfoStats.kt [hunk 2]
```diff
@@ -32,17 +32,38 @@ class CivInfoStats(val civInfo: CivilizationInfo) {
...
+            }else{
+                unit.maintenance = 1f
+            }
+        }
+        // Apply global discounts
+        for (unique in civInfo.getMatchingUniques(UniqueType.UnitMaintenanceDiscountGlobal, StateForConditionals(civInfo))) {
+            for (unit in unitsToPayFor.filter{it.matchesFilter(unique.params[1])}){
+                unit.maintenance *= 1f + unique.params[0].toFloat()/100f
```
Comment (yairm210) [inline core/src/com/unciv/logic/civilization/CivInfoStats.kt:50]: And here

File: core/src/com/unciv/logic/civilization/CivInfoStats.kt [hunk 3]
```diff
@@ -32,17 +32,38 @@ class CivInfoStats(val civInfo: CivilizationInfo) {
...
-                numberOfUnitsToPayFor,
-                unitsToPayFor.count { it.matchesFilter(unique.params[1]) }.toFloat()
-            )
-            numberOfUnitsToPayFor += numberOfUnitsWithDiscount * unique.params[0].toFloat() / 100f
+        // Handle unit maintenance discounts
+        // Have to capture global and per-unit
+        // Free Garrison already removed above from sequence
+        // Initialize maintenance cost per unit, default 1 aka 100%
+        // Note all discounts are in the form of -X%, such as -25 for 25% reduction
+        for (unit in unitsToPayFor){
```
Comment (yairm210) [inline core/src/com/unciv/logic/civilization/CivInfoStats.kt:40]: Foreach unit
   Maintenance=1
   For(unique in getMatchingUniques(..)) maintenance *= unique.params[]. toPercent ()

File: core/src/com/unciv/logic/civilization/CivInfoStats.kt [hunk 4]
```diff
@@ -32,17 +32,38 @@ class CivInfoStats(val civInfo: CivilizationInfo) {
...
+            for (unit in unitsToPayFor.filter{it.matchesFilter(unique.params[1])}){
+                unit.maintenance *= 1f + unique.params[0].toFloat()/100f
+            }
         }
+        // Sort by descending maintenance, then drop most expensive X units to make them free
+        // If more free than units left, returns empty sequence
+        unitsToPayFor = unitsToPayFor.sortedByDescending { it.maintenance }.drop(freeUnits)
+        val numberOfUnitsToPayFor = max(0.0, unitsToPayFor.sumOf { it.maintenance.toDouble() }).toFloat()
+
+//        var numberOfUnitsToPayFor = max(0f, unitsToPayFor.count().toFloat() - freeUnits)
```
Comment (yairm210) [inline core/src/com/unciv/logic/civilization/CivInfoStats.kt:58]: Remove dead code

File: android/assets/jsons/Civ V - Gods & Kings/Units.json [hunk 5]
```diff
@@ -1474,6 +1474,7 @@
 		"range": 8,
 		"cost": 150,
 		"requiredTech": "Advanced Ballistics",
+		"uniques": ["-100% maintenance costs"],
```
Comment (yairm210) [inline android/assets/jsons/Civ V - Gods & Kings/Units.json:1477]: Number needs to be in square brackets

**Comment by itanasi** [comment]
Thanks for the comments. Code cleaned up and this time validated against actual commit

**Pull request merged**